core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20210615_sharded_entity_tables.xml

```diff
@@ -0,0 +1,271 @@
+            DROP PROCEDURE IF EXISTS createEntityAttributeShard /
+            CREATE PROCEDURE createEntityAttributeShard(IN shardIdentifier CHAR(4))
+            BEGIN
+                SET @tableSuffix := shardIdentifier;
+                SET @sql_text:=CONCAT('CREATE TABLE ENTITY_ATTRIBUTE_',@tableSuffix,' ('
+                    ' id bigint(20) unsigned NOT NULL AUTO_INCREMENT,'
+                    ' name varchar(200) NOT NULL,'
+                    ' value_string text,'
+                    ' value_number double DEFAULT NULL,'
+                    ' value_boolean bit(1) DEFAULT NULL,'
+                    ' value_entity_ref bigint(20) unsigned DEFAULT NULL,'
+                    ' list_index int(11) DEFAULT NULL,'
+                    ' owner_id bigint(20) unsigned NOT NULL,'
+                    ' list_length int(11) DEFAULT NULL,'
+                    ' namespace varchar(32) NOT NULL DEFAULT ''default'','
+                    ' VALUE_JSON longtext,'
+                    ' deleted bit(1) DEFAULT b''0'','
+                    ' deleted_date timestamp(6) NULL DEFAULT NULL,'
+                    ' PRIMARY KEY (id),'
+                    ' KEY FK_ENT_ATTRIBUTE_ENTITY_REF (value_entity_ref),'
+                    ' KEY UNQ_ENTITY_ATTRIBUTE (owner_id,namespace,name,list_index),'
+                    ' CONSTRAINT FK_ATTRIBUTE_PARENT_ENTITY_',@tableSuffix,' FOREIGN KEY (owner_id) REFERENCES ENTITY (id),'
+                    ' CONSTRAINT FK_ENT_ATTRIBUTE_ENTITY_REF_',@tableSuffix,' FOREIGN KEY (value_entity_ref) REFERENCES ENTITY (id)'
+                    ') ENGINE=InnoDB AUTO_INCREMENT=231805857 DEFAULT CHARSET=utf8;');
+                PREPARE stmt from @sql_text;
+                EXECUTE stmt;
+            END /
+
+            -- procedure to copy rows from ENTITY_ATTRIBUTE to the appropriate shard table,
+            -- based on the attributes' owning workspace
+            DROP PROCEDURE IF EXISTS copyRowsToAttributeShard /
+            CREATE PROCEDURE copyRowsToAttributeShard(IN shardIdentifier CHAR(4))
+            BEGIN
+                SET @tableSuffix := shardIdentifier;
+                SET @sql_text:=CONCAT('INSERT INTO ENTITY_ATTRIBUTE_',@tableSuffix,'(id, name, value_string, value_number, value_boolean,'
+                    ' value_entity_ref, list_index, owner_id, list_length, namespace, VALUE_JSON, deleted, deleted_date)'
+                    'SELECT ea.id, ea.name, ea.value_string, ea.value_number, ea.value_boolean,'
+                    ' ea.value_entity_ref, ea.list_index, ea.owner_id, ea.list_length, ea.namespace, ea.VALUE_JSON, ea.deleted, ea.deleted_date'
+                    ' FROM ENTITY_ATTRIBUTE ea, ENTITY e, WORKSPACE w'
+                    ' WHERE e.workspace_id = w.id'
+                    ' AND ea.owner_id = e.id'
+                    ' AND shardIdentifier(hex(w.id)) = ''',@tableSuffix,''';');
+                PREPARE stmt from @sql_text;
+                EXECUTE stmt;
```

Comment:
TODO: disable unique/foreign key checks and defer indexes until after the insert is done?

Reply:
The command line tools for MySQL and Postgres all seem to do this.  I think we are going to be surprised by how long it takes to conduct operations on 10s and 100s of millions of records so my guess it's probably worth the extra code in terms of the time it will save us.